orchestration/src/main/java/com/sap/ai/sdk/orchestration/PolymorphicFallbackDeserializer.java

```diff
@@ -0,0 +1,105 @@
+package com.sap.ai.sdk.orchestration;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.google.common.annotations.Beta;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nonnull;
+import lombok.AllArgsConstructor;
+
+/**
+ * Handles polymorphic deserialization for a base class or interface.
+ *
+ * <p>This deserializer uses a fallback strategy by attempting to deserialize the JSON into known
+ * subtypes. Subtypes are either discovered using the {@link JsonSubTypes} annotation or provided
+ * explicitly. If deserialization fails for all candidates, a {@link JsonMappingException} is thrown
+ * with suppressed exceptions.
+ *
+ * @since 1.2.0
+ * @param <T> The base type for deserialization.
+ */
+@Beta
+@AllArgsConstructor(access = lombok.AccessLevel.PRIVATE)
```

Comment:
(Minor)

```suggestion
@AllArgsConstructor(access = lombok.AccessLevel.PROTECTED)
```

No one can instantiate this class outside of the namespace. However this might be necessary for user workarounds. Could you change it to `protected`? Then extending will be possible.

Reply:
I was under the impression, we go for the 'strictest' and relax it as and when needed. 

Also, is there a scenario where `protected` static factory methods is not enough and a `public` constructor needs to be exposed?

Reply:
> I was under the impression, we go for the 'strictest' and relax it as and when needed.

I think we are not fully aligned in the team on this.
I understand strictest is safest.
But personally after so many years troubleshooting, debugging and figuring out workarounds for users who struggle with some edge case (we haven't thought of): It's convenient to offer a workaround solution with protected API, than not being able to help at all or letting the user wait for another release cycle to open up the API a little.

`protected` is considered semi-public. Since it's also annotated as `@Beta`, I see no danger in potentially changing it in the future.

> Also, is there a scenario where `protected` static factory methods are not enough and a `protected` constructor needs to be exposed?

I haven't tested it myself. But if you find a way to use these `protected` static methods from outside of the namespace, then we don't need the constructor - true! I think it's only possible via extending the class, re-defining a custom static method to call `super.[...]` static methods. This would look rather ugly, even for a workaround.

Reply:
Thanks for the insight! :)